Clarify OnAckReceived pseudocode (#2140)
quicwg/base-drafts: draft-ietf-quic-recovery.md

: The packet number of the most recently sent packet.

largest_acked_packet:

> : The largest packet number acknowledged by any ACK frame.

Reviewer: ```suggestion
: The largest packet number acknowledged in the connection so far.
```

Reviewer 2: Isn't this in the packet number space?

Reviewer 2: I think that you mean "packet number space" rather than connection.

~~~
  OnAckReceived(ack):

>     largest_acked_packet = max(largest_acked_packet,

Reviewer 3: I think this accurately describes the text we have, but I think we should add some text that you MAY use the largest acked from the most recent ACK, which could reduce spurious retransmits when one packet arrives far out of order?

Reviewer 3: To clarify, I'm ok doing this in a follow up PR.

Author: > you MAY use the largest acked from the most recent ACK, which could reduce spurious retransmits when one packet arrives far out of order

Any reason not to make this the sole recommendation? I had some trouble coming up with a case where that might occur, which is why I didn't take that route originally, but I think I see it now: when a reordered ACK newly acknowledges a packet such that the newly acknowledged packet is > ack.largest_acked - kPacketThreshold, but <= largest_acked_packet - kPacketThreshold.

Author: Er, wait, that's nonsense. Still having trouble seeing the case.

Reviewer 3: It's the case when one packet is far out of order, and then subsequent packets arrive in order.  ie: 100 is received, then 1, 2, 3, 4, 5, etc.  I don't fully understand the cause, but I've seen small UDP packets jump ahead of large ones in a number of traces.

But I suspect Jana and I will merge this as is and we can add follow-up text about that issue if we think it's valuable.

    // ack-eliciting, update the RTT.
    if (sent_packets[ack.largest_acked] &&
        sent_packets[ack.largest_acked].ack_eliciting):

>       largest_acked_packet = ack.largest_acked

Reviewer 2: This doesn't work because the largest_acked in the ACK might be newly acknowledged, but not the largest ever acknowledged.

```suggestion
      if ack.largest_acked > largest_acked_packet:
         largest_acked_packet = ack.largest_acked
```

      latest_rtt = now - sent_packets[ack.largest_acked].time
      UpdateRtt(latest_rtt, ack.ack_delay)

    // Process ECN information if present.
    if (ACK frame contains ECN information):

>        ProcessECN(ack)

Reviewer 2: This is a separate change.  I think that you should make this a separate PR.
